### PR TITLE
Streamhistory

### PR DESCRIPTION
**Pull request by Readon**
Closes #

# Context, Motivation & Description

StreamHistory is a kind of history which would have a valid for each element to demonstrate if the data is valid for using.
It acts like FIFO, also follow first in first out policy, but with internal data visibility.

# Impact on code generation

# Checklist

- [ x ] Unit tests were added
- [ ] API changes are or will be documented:
  - using Scaladoc comments: `/** */`?
  - on [RTD](https://github.com/SpinalHDL/SpinalDoc-RTD)?
  - thanks to a [new tracking issue on RTD](https://github.com/SpinalHDL/SpinalDoc-RTD/issues/new?title=Document%20XXX&body=Do%20not%20merge%20until%20SpinalHDL/SpinalHDL%23XXX%20has%20not%20been%20merged)?

### CODE REVIEW

File: lib/src/main/scala/spinal/lib/Stream.scala [hunk 1]
```diff
@@ -1627,6 +1627,50 @@ class StreamFifoCC[T <: Data](val dataType: HardType[T],
...
+  def builder(prev: Stream[T], left: Int): List[Stream[T]] = {
+    left match {
+      case 0 => Nil
+      case 1 => prev :: Nil
+      case _ => prev :: builder(
+        {
+          val stream = Stream(dataType)
+          stream <-< prev
+
+          val next = Stream(dataType)
```
Comment (Dolu1990) [inline lib/src/main/scala/spinal/lib/Stream.scala:1657]: I'm not sure to understand why "next" stream is required, why not returning directly "stream" ?

Comment (Readon) [reply lib/src/main/scala/spinal/lib/Stream.scala:1657]: the stream's ready should be also asserted when it's valid is not true.
next is used for interface only.

File: lib/src/main/scala/spinal/lib/Stream.scala [hunk 2]
```diff
@@ -1627,6 +1627,50 @@ class StreamFifoCC[T <: Data](val dataType: HardType[T],
...
+          next.valid := stream.valid
+          next.payload := stream.payload
+          stream.ready := next.ready | !stream.valid
```
Comment (Readon) [inline lib/src/main/scala/spinal/lib/Stream.scala:1660]: @Dolu1990 Here is the point.

File: lib/src/main/scala/spinal/lib/Stream.scala (hunk 1 above)
Comment (Readon) [reply lib/src/main/scala/spinal/lib/Stream.scala:1657]: that means elements pushed at input would automatically move forward to pop end.

Comment (Dolu1990) [reply lib/src/main/scala/spinal/lib/Stream.scala:1657]: > the stream's ready should be also asserted when it's valid is not true.

The <-< operator does that already by default i think (it collapse bubbles) : 
https://github.com/SpinalHDL/SpinalHDL/blob/c150e90db33e389503821077caa942194745b271/lib/src/main/scala/spinal/lib/Stream.scala#L369

Also, as by default it collapse bubbles, the ready signal has no meaning when the valid is low.

or i missed something ?

Comment (Readon) [reply lib/src/main/scala/spinal/lib/Stream.scala]: You are right, I missed the collapse bubbles function previously.
Now it has been fixed.

**Pull request merged**